news/tests/test_tasks.py

```diff
@@ -232,3 +233,143 @@ def myfunc():
+        self.assertEqual(record['FIRST_DEVICE'], 'y')
+
+        record = self._base_test(first_device=False)
+        self.assertEqual(record['FIRST_DEVICE'], 'n')
+
+    def test_windows(self):
+        ua = 'Mozilla/5.0 (Windows NT 6.1; rv:10.0) Gecko/20100101 Firefox/10.0'
+        record = self._base_test(ua)
+        self.assertEqual(record['OS'], 'Windows 7')
+        self.assertEqual(record['OS_VERSION'], '')  #TODO Not sure if we expect '7' here.
+        self.assertEqual(record['BROWSER'], 'Firefox 10')
+        self.assertEqual(record['DEVICE_NAME'], 'Other')
+        self.assertEqual(record['DEVICE_TYPE'], 'D')
+
+    def test_mac(self):
+        ua = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10.6; rv:10.0) Gecko/20100101 Firefox/30.2'
+        record = self._base_test(ua)
+        self.assertEqual(record['OS'], 'Mac OS X')
+        self.assertEqual(record['OS_VERSION'], '10.6')
+        self.assertEqual(record['BROWSER'], 'Firefox 30.2')
+        self.assertEqual(record['DEVICE_NAME'], 'Other')
+        self.assertEqual(record['DEVICE_TYPE'], 'D')
+
+    def test_linux(self):
+        ua = 'Mozilla/5.0 (X11; Linux i686 on x86_64; rv:10.0) Gecko/20100101 Firefox/42.0'
+        record = self._base_test(ua)
+        self.assertEqual(record['OS'], 'Linux')
+        self.assertEqual(record['OS_VERSION'], '')
+        self.assertEqual(record['BROWSER'], 'Firefox 42')
+        self.assertEqual(record['DEVICE_NAME'], 'Other')
+        self.assertEqual(record['DEVICE_TYPE'], 'D')
+
+    def test_android_phone_below_version_41(self):
+        ua = 'Mozilla/5.0 (Android; Mobile; rv:40.0) Gecko/40.0 Firefox/40.0'
+        record = self._base_test(ua)
+        self.assertEqual(record['OS'], 'Android')
+        self.assertEqual(record['OS_VERSION'], '')
+        self.assertEqual(record['BROWSER'], 'Firefox Mobile 40')
+        self.assertEqual(record['DEVICE_NAME'], 'Generic Smartphone')
+        self.assertEqual(record['DEVICE_TYPE'], 'M')
+
+    def test_android_tablet_below_version_41(self):
+        ua = 'Mozilla/5.0 (Android; Tablet; rv:40.0) Gecko/40.0 Firefox/40.0'
+        record = self._base_test(ua)
+        self.assertEqual(record['OS'], 'Android')
+        self.assertEqual(record['OS_VERSION'], '')
+        self.assertEqual(record['BROWSER'], 'Firefox 40')
+        self.assertEqual(record['DEVICE_NAME'], 'Generic Tablet')
+        self.assertEqual(record['DEVICE_TYPE'], 'T')
+
+    def test_android_phone_from_version_41(self):
+        ua = 'Mozilla/5.0 (Android 4.4; Mobile; rv:41.0) Gecko/41.0 Firefox/41.0'
+        record = self._base_test(ua)
+        self.assertEqual(record['OS'], 'Android')
+        self.assertEqual(record['OS_VERSION'], '4.4')
+        self.assertEqual(record['BROWSER'], 'Firefox Mobile 41')
+        self.assertEqual(record['DEVICE_NAME'], 'Generic Smartphone')
+        self.assertEqual(record['DEVICE_TYPE'], 'M')
+
+    # TODO This reports Android 5 instead of Firefox 40
```

Comment:
I didn't manage to get this working. `ua_parse` lib seems to get this wrong for both chrome and firefox. Filed bug https://github.com/ua-parser/uap-core/issues/77